jps-plugin/src/org/jetbrains/jps/incremental/scala/data/CompilerData.scala

```diff
@@ -27,8 +27,15 @@ object CompilerData {
     val compilerJars = if (SettingsManager.hasScalaSdk(module)) {
       compilerJarsIn(module).flatMap { case jars: CompilerJars =>
         val absentJars = jars.files.filter(!_.exists)
+        val compileJars =
+          if (absentJars.isEmpty && SettingsManager.getHydraSettings(project.getProject).isHydraEnabled
+            && SettingsManager.getHydraSettings(project.getProject).getArtifactPaths.containsKey(compilerVersion(module).getOrElse("UNKNOWN"))) {
```

Comment:
I'd also refactor a bit the code so that it's easier to read (the condition is really long and I'd prefer to have a method with an intention revealing name), e.g.,

```
def useHydraCompiler(jars: CompilerJars): Boolean = {
  val missingJars = jars.files.filter(!_.exists)
  // LOG if JARS are missing
  val hydraProjectSettings = SettingsManager.getHydraSettings(project.getProject)
  missingJars.isEmpty && hydraProjectSettings.isHydraEnabled && compilerVersion(module).nonEmpty && hydraProjectSettings.getArtifactPaths.containsKey(compilerVersion(module).get)
}

val compileJars = {
  if (useHydraCompiler(jars)) {
    // logic
  }
  else jars
}
```

By the way, can `hydraProjectSettings.getArtifactPaths` actually return "UNKNOWN"? (The above refactored implementation assumes that's never the case)

Reply:
No, `hydraProjectSettings.getArtifactPaths` can't return "UNKNOWN".

Reply:
Alright, then it's better not to have "UNKNOWN" in the code.